### PR TITLE
[FIXED JENKINS-25416][JENKINS-28790] Do not inject build variables into maven process optionally

### PR DESCRIPTION
**Pull request by olivergondza**
Fixing [JENKINS-26684](https://issues.jenkins-ci.org/browse/JENKINS-26684) exposed several problems, where at least [JENKINS-25416](https://issues.jenkins-ci.org/browse/JENKINS-25416) does not seem fixable.

The maven build step injects all build variables into maven process causing different kind of issues. As there are a lot of build relying on this behaviour, I do not think this default can be changed. Instead, this provides an option to turn the feature off in case it is causing problems.

Can be used to workaround [JENKINS-28790](https://issues.jenkins-ci.org/browse/JENKINS-28790) as well.

### CODE REVIEW

**Comment by KostyaSha** [comment]
:+1: that same issues exist in all plugins that copied maven, afair ant and probably gradle.

File: core/src/main/java/hudson/tasks/Maven.java [hunk 1]
```diff
@@ -155,8 +161,12 @@ public Maven(String targets,String name, String pom, String properties, String j
         this(targets, name, pom, properties, jvmOptions, usePrivateRepository, null, null);
     }
     
-    @DataBoundConstructor
     public Maven(String targets,String name, String pom, String properties, String jvmOptions, boolean usePrivateRepository, SettingsProvider settings, GlobalSettingsProvider globalSettings) {
+        this(targets, name, pom, properties, jvmOptions, usePrivateRepository, settings, globalSettings, true);
+    }
+
+    @DataBoundConstructor
+    public Maven(String targets,String name, String pom, String properties, String jvmOptions, boolean usePrivateRepository, SettingsProvider settings, GlobalSettingsProvider globalSettings, boolean injectBuildVariables) {
```
Comment (KostyaSha) [inline core/src/main/java/hudson/tasks/Maven.java]: Use `DataBoundSetter`

File: core/src/main/java/hudson/tasks/Maven.java [hunk 2]
```diff
@@ -137,6 +137,12 @@
      */
     private GlobalSettingsProvider globalSettings;
 
+    /**
+     * Inject build variables as properties into maven process.
+     * @since TODO
+     */
+    private boolean injectBuildVariables = true;
```
Comment (jtnord) [inline core/src/main/java/hudson/tasks/Maven.java]: @KostyaSha yes - but restricted to buildVars as systemProperties - which I think may need a little more thightening of the name - but that's nit picking on my part!

Comment (jtnord) [inline core/src/main/java/hudson/tasks/Maven.java]: taking a *quick* peak at the code it looks like we should already set environment variables (unless I missed something)

Comment (KostyaSha) [inline core/src/main/java/hudson/tasks/Maven.java]: Smells like test case.

**Comment by olivergondza** [comment]
@jtnord, I am afraid you are missing the point. The injection was performed unconditionally for years and changing (even the default to false) would break a lot of builds that relies on properties. How much I like to get rid of this *feature*, I do not think we want to make this change.

**Comment by KostyaSha** [comment]
@olivergondza i think @jtnord means exclude injection, but keep ability to pass `-Dpar=${env.variable}` because `env` is probably not a standard variable and exists only in maven-plugin.

File: core/src/main/java/hudson/tasks/Maven.java (hunk 2 above)
Comment (olivergondza) [inline core/src/main/java/hudson/tasks/Maven.java]: @jtnord That should be done for all launched processes if I am not mistaken.

File: core/src/main/java/hudson/tasks/Maven.java [hunk 3]
```diff
@@ -137,6 +137,12 @@
      */
     private GlobalSettingsProvider globalSettings;
 
+    /**
+     * Inject build variables as properties into maven process.
```
Comment (KostyaSha) [inline core/src/main/java/hudson/tasks/Maven.java]: I think you need add `ALL build vars` as it the main disaster. And note that escaped vars may cause issues.

**Comment by jtnord** [comment]
@olivergondza What I am saying is that the default for **new** jobs should be false (i.e. you opt in to getting your process broken) - to gently prod users to either enable the "I accept this may break my jobs" or us the alternative `env.xxx` form - not that it should be flipped for existing jobs.

In order to do this you *may* need a `Boolean` rather than a `boolean` - but that should be possible.  Using the Boolean approach people pushing config.xml or even generating it by hand should be unaffected - so we are talking about users creating this from the UI only (which I have never seen automated for generating jobs).  This gives existing users and jobs time to migrate - and in Jenkins 55.0 we can remove the *feature*

**Comment by olivergondza** [comment]
Feature introduced in https://github.com/jenkinsci/jenkins/commit/d2bb200c4e8a31f0541466d99110dfa38425185e

**Comment by olivergondza** [comment]
> @olivergondza What I am saying is that the default for new jobs should be false (i.e. you opt in to getting your process broken) - to gently prod users to either enable the "I accept this may break my jobs" or us the alternative  env.xxx  form - not that it should be flipped for existing jobs.

I see. Let me rework that.

**Comment by KostyaSha** [comment]
Now unmergeable

File: core/src/main/java/hudson/tasks/Maven.java (hunk 3 above)
Comment (olivergondza) [inline core/src/main/java/hudson/tasks/Maven.java]: I am not sure this is relevant on javadoc level.

**Comment by olivergondza** [comment]
Thanks guys, I have implemented @jtnord's suggestion using the negative field trick suggested by @KostyaSha.

File: core/src/main/java/hudson/tasks/Maven.java [hunk 4]
```diff
@@ -137,6 +141,15 @@
...
     private GlobalSettingsProvider globalSettings;
 
+    /**
+     * Skip injecting build variables as properties into maven process.
+     *
+     * Defaults to false to mimic the legacy behavior.
+     *
+     * @since TODO
+     */
+    private boolean doNotInjectBuildVariables = false;
```
Comment (KostyaSha) [inline core/src/main/java/hudson/tasks/Maven.java]: maybe name shorter i.e. `noInjectedVars`?

Comment (olivergondza) [inline core/src/main/java/hudson/tasks/Maven.java]: I used the most descriptive name I can think off, that is the only think that matters for a field name IMO.

File: core/src/main/java/hudson/tasks/Maven.java [hunk 5]
```diff
@@ -155,8 +168,12 @@ public Maven(String targets,String name, String pom, String properties, String j
         this(targets, name, pom, properties, jvmOptions, usePrivateRepository, null, null);
     }
     
-    @DataBoundConstructor
     public Maven(String targets,String name, String pom, String properties, String jvmOptions, boolean usePrivateRepository, SettingsProvider settings, GlobalSettingsProvider globalSettings) {
+        this(targets, name, pom, properties, jvmOptions, usePrivateRepository, settings, globalSettings, true);
```
Comment (jtnord) [inline core/src/main/java/hudson/tasks/Maven.java]: @deprecated unless you are going to use a `DataBoundSetter` as suggested by @KostyaSha

Comment (KostyaSha) [inline core/src/main/java/hudson/tasks/Maven.java]: or make setter and use it from this constructor, then there will be 2 constructors: 1 with default, 2 with setting

File: core/src/main/resources/hudson/tasks/Maven/help-injectBuildVariables.html [hunk 6]
```diff
@@ -0,0 +1,3 @@
+<div>
+  Pass all build variables into maven process in form of java properties. This is seldom needed as Jenkins provides it as environment variables anyway.
```
Comment (jtnord) [inline core/src/main/resources/hudson/tasks/Maven/help-injectBuildVariables.html]: and specific variables can be passed by appending `-Dmyvariable=$SOME_BUILD_VAR` to the maven goals?

(at least I believe this works)

Comment (KostyaSha) [inline core/src/main/resources/hudson/tasks/Maven/help-injectBuildVariables.html]: teeeest case it!

Comment (olivergondza) [inline core/src/main/resources/hudson/tasks/Maven/help-injectBuildVariables.html]: @jtnord, it works but the preferred way is to use the properties textarea (covered by `parametersReferencedFromPropertiesShouldRetainBackslashes` at least).

@KostyaSha, it is not a shame to contribute PR with nothing but tests, you know :wink:.

**Comment by jtnord** [comment]
well I am happy that this is an improvement over the existing code so :+1:

File: core/src/main/java/hudson/tasks/Maven.java (hunk 5 above)
Comment (olivergondza) [inline core/src/main/java/hudson/tasks/Maven.java]: I prefer to leave it as it is. None of the old constructors that got "extended" was deprecated.

Ad `@DataBoundSetter`: There are 9 fields in the class:

- 5 of them is final (some have defaults set in constructor overloads)
- 3 can be set by constructor and updated by setter (one of them directly!)
- if we use `@DataBoundSetter` there will be one filed that can not be set by constructor but only using setter.

I am trying not to make it any more messy then it already is.

**Comment by oleg-nenashev** [comment]
👍 from me. But there is a merge conflict again 😞

**Comment by olivergondza** [comment]
Thanks for the reviews. Let's merge it if tests are green.

**Comment by daniel-beck** [comment]
Wouldn't it be better to switch the default? I.e. existing build steps get `readResolve`d to pass these system properties as arguments, but new build steps don't?

**Comment by olivergondza** [comment]
@daniel-beck, that is how it works. Old configs will get the old behaviour (inject everything) and users will have an option to turn this off (skip injecting). New configs will have this off unless users opt-in.

**Comment by daniel-beck** [comment]
@olivergondza I see. You went for the inverted meaning property rather than `readResolve`ing to `true`.

I wonder whether that's an issue in a later move to `@DataboundSetter`, which is more Pipeline-friendly.

**Comment by olivergondza** [comment]
> I wonder whether that's an issue in a later move to @DataboundSetter, which is more Pipeline-friendly.

The inverted meaning is used only internally so UI and Pipeline definition should be unified. I see no problem with `@DataBoundSetter` inverting the flag, when we decide to go that way.

**Comment by daniel-beck** [comment]
@olivergondza The advantage of `@DataboundSetter` is to not have to set every conceivable property explicitly (as in `@DataboundConstructor`) when using `step` in Pipeline. However, when not setting this property, it will default to `false`, which is the legacy behavior. Or am I missing something here?

**Comment by olivergondza** [comment]
I do not feel strongly here. Refactored to avoid the negation approach to be less surprising.

---

> However, when not setting this property, it will default to false, which is the legacy behavior. Or am I missing something here?

All the constructors either accepted parameter or set it to true so it is not the case. Yes, the initializer was probably useless in all code paths.

**Comment by oleg-nenashev** [comment]
@daniel-beck @olivergondza Did you agree regarding the merge?

**Comment by daniel-beck** [comment]
LGTM 👍

**Comment by oleg-nenashev** [comment]
Seems the last PR build has never finished. Actually it lead to the CI breakage: https://jenkins.ci.cloudbees.com/job/core/job/jenkins-core/5557/testReport/hudson.tasks/MavenTest/parametersReferencedFromPropertiesShouldRetainBackslashes/

The test injects PATH, so the behavior changed after this PR. https://github.com/jenkinsci/jenkins/blame/e66e719f1e6caa54b546a5bdee5ab062089d21c2/test/src/test/java/hudson/tasks/MavenTest.java#L221

Needs a fix

**Pull request merged**